File: src/net.mjs
// Networking (SPEC §8). fetchHtml over undici's fetch (redirects + gzip/br/deflate
// decode for free), plus the hardening the spec calls for: charset sniffing, max
// body size, content-type gate, and an optional CookieJar round-trip. We use the
// undici package's own fetch (not Node's global) so it pairs with the undici
// Agent we pass as a `dispatcher` for HTTP/2 — mixing the two undici builds
// throws "invalid onRequestStart method".

import { fetch as undiciFetch } from "undici";

const DEFAULT_UA = "turbo-crawl/0.0 (+https://github.com/miaskiewicz/turbo-crawl)";
const DEFAULT_MAX_BYTES = 8 * 1024 * 1024; // 8 MiB
const HTML_TYPES = ["text/html", "application/xhtml+xml", "application/xml", "text/xml"];

export class HttpError extends Error {
  constructor(message, code) {
    super(message);
    this.name = "HttpError";
    this.code = code;
  }
}

// charset: Content-Type header → <meta charset> sniff over the first bytes → utf-8.
function detectCharset(headers, head) {
  const ct = headers.get("content-type") ?? "";
  const m = /charset=([^;]+)/i.exec(ct);
  if (m) return m[1].trim().replace(/["']/g, "").toLowerCase();
  const ascii = new TextDecoder("latin1").decode(head);
  const meta =
    /<meta[^>]+charset=["']?([\w-]+)/i.exec(ascii) ||
    /<meta[^>]+content=["'][^"']*charset=([\w-]+)/i.exec(ascii);
  return meta ? meta[1].toLowerCase() : "utf-8";
}

function isHtmlType(headers) {
  const ct = (headers.get("content-type") ?? "").toLowerCase();
  if (!ct) return true; // be permissive when the server omits it
  return HTML_TYPES.some((t) => ct.includes(t));
}

// Reject up-front when the declared content-length already blows the cap.
function checkContentLength(res, maxBytes) {
  const len = Number(res.headers.get("content-length"));
  if (len && len > maxBytes) {
    throw new HttpError(`body exceeds maxBytes (${len} > ${maxBytes})`, "BODY_TOO_LARGE");
  }
}

// Flatten accumulated chunks into one contiguous Uint8Array.
function concatChunks(chunks, total) {
  const out = new Uint8Array(total);
  let off = 0;
  for (const c of chunks) {
    out.set(c, off);
    off += c.length;
  }
  return out;
}

// Stream the body, enforcing the cap as bytes arrive.
async function accumulate(body, maxBytes) {
  const chunks = [];
  let total = 0;
  for await (const chunk of body) {
    total += chunk.length;
    if (total > maxBytes) {
      throw new HttpError(`body exceeds maxBytes (> ${maxBytes})`, "BODY_TOO_LARGE");
    }
    chunks.push(chunk);
  }
  return concatChunks(chunks, total);
}

// Read the body with a hard byte cap, streaming so we never buffer a huge response.
async function readCapped(res, maxBytes) {
  checkContentLength(res, maxBytes);
  if (!res.body) return new Uint8Array(0);
  return accumulate(res.body, maxBytes);
}

function decode(bytes, charset) {
  try {
    return new TextDecoder(charset).decode(bytes);
  } catch {
    return new TextDecoder("utf-8").decode(bytes); // unknown label → utf-8
  }
}

/**
 * Fetch a URL and return its decoded HTML plus response metadata.
 *
 * @param {string} url
 * @param {object} [opts]
 * @param {Record<string,string>} [opts.headers]
 * @param {string} [opts.method]
 * @param {string} [opts.body]
 * @param {AbortSignal} [opts.signal]
 * @param {import('./cookies.mjs').CookieJar} [opts.jar]  request/response cookie round-trip
 * @param {object} [opts.cache]      ResponseCache for conditional (304) revalidation
 * @param {object} [opts.dispatcher] undici Agent (HTTP/2 + DNS cache) as the fetch dispatcher
 * @param {number} [opts.maxBytes]
 * @param {boolean} [opts.allowNonHtml]  skip the content-type gate
 * @param {typeof fetch} [opts.fetch]    injectable for tests / Lane B
 * @returns {Promise<{ html:string, finalUrl:string, status:number, headers:Headers }>}
 */
// Build the request headers: UA/accept defaults, caller overrides, jar cookie,
// and conditional-request validators (If-None-Match / If-Modified-Since).
function buildHeaders(url, opts) {
  const headers = {
    "user-agent": DEFAULT_UA,
    accept: "text/html,application/xhtml+xml",
    ...opts.headers,
  };
  if (opts.jar) {
    const cookie = opts.jar.cookieHeader(url);
    if (cookie) headers.cookie = cookie;
  }
  if (opts.cache) Object.assign(headers, opts.cache.validators(url));
  return headers;
}

// Feed any Set-Cookie headers from the response back into the jar.
function ingestSetCookie(opts, res, finalUrl) {
  if (opts.jar && typeof res.headers.getSetCookie === "function") {
    opts.jar.setFromResponse(finalUrl, res.headers.getSetCookie());
  }
}

// Enforce the HTML content-type gate unless the caller opted out.
function gateHtmlType(opts, res) {
  if (!opts.allowNonHtml && !isHtmlType(res.headers)) {
    throw new HttpError(`non-HTML content-type: ${res.headers.get("content-type")}`, "NOT_HTML");
  }
}

// Read + charset-decode the body into the final HTML string.
async function decodeBody(res, maxBytes) {
  const bytes = await readCapped(res, maxBytes);
  return decode(bytes, detectCharset(res.headers, bytes.subarray(0, 1024)));
}

const REDIRECT_STATUS = new Set([301, 302, 303, 307, 308]);

function redirectLocation(res) {
  return REDIRECT_STATUS.has(res.status) ? res.headers.get("location") : null;
}

// Method after a redirect: 303 (and 301/302 on POST) become GET; 307/308 keep it.
function nextMethod(status, method) {
  if (status === 303) return "GET";
  if ((status === 301 || status === 302) && method === "POST") return "GET";
  return method;
}

// Compute the next hop's state after a redirect (URL + rewritten method/body).
function advanceHop(state, res, loc) {
  const method = nextMethod(res.status, state.method);
  return {
    current: new URL(loc, state.current).href,
    method,
    body: method === state.method ? state.body : undefined,
  };
}

// One request in the manual-follow loop: returns { done } at a terminal response
// (or the cap) else { next } with the following hop's state.
async function redirectHop(doFetch, state, opts, redirects, maxRedirects) {
  const headers = buildHeaders(state.current, { ...opts, method: state.method });
  const res = await doFetch(state.current, {
    method: state.method,
    body: state.body,
    redirect: "manual",
    signal: opts.signal,
    headers,
    dispatcher: opts.dispatcher,
  });
  ingestSetCookie(opts, res, res.url || state.current);
  const loc = redirectLocation(res);
  if (!loc || redirects >= maxRedirects) {
    return { done: { res, finalUrl: res.url || state.current, redirected: redirects > 0 } };
  }
  return { next: advanceHop(state, res, loc) };
}

// Manual redirect following with a hard hop cap (opts.maxRedirects). Re-derives
// the Cookie header + Set-Cookie ingest per hop and rewrites method/body per the
// fetch spec. Used only when the caller sets maxRedirects.
async function followManually(doFetch, url, opts, maxRedirects) {
  let state = { current: url, method: opts.method ?? "GET", body: opts.body };
  for (let redirects = 0; ; redirects++) {
    const step = await redirectHop(doFetch, state, opts, redirects, maxRedirects);
    if (step.done) return step.done;
    state = step.next;
  }
}

// Single fetch delegating redirects to undici (default: cap 20).
async function followAuto(doFetch, url, opts) {
  const res = await doFetch(url, {
    method: opts.method ?? "GET",
    body: opts.body,
    redirect: "follow",
    signal: opts.signal,
    headers: buildHeaders(url, opts),
    dispatcher: opts.dispatcher,
  });
  const finalUrl = res.url || url;
  ingestSetCookie(opts, res, finalUrl);
  return { res, finalUrl, redirected: !!res.redirected };
}

// 304 Not Modified → reuse the cached body (the server sent no body).
function notModified(res, finalUrl, redirected, opts) {
  return {
    html: opts.cache.body(finalUrl),
    finalUrl,
    status: 304,
    headers: res.headers,
    redirected,
    notModified: true,
  };
}

// Turn a settled response into the FetchResult: short-circuit 304s, else gate +
// decode and (when caching) record the validators/body for next time.
async function finishFetch(res, finalUrl, redirected, opts, maxBytes) {
  if (opts.cache && res.status === 304) return notModified(res, finalUrl, redirected, opts);
  gateHtmlType(opts, res);
  const html = await decodeBody(res, maxBytes);
  if (opts.cache) opts.cache.store(finalUrl, res.headers, html, res.status);
  return { html, finalUrl, status: res.status, headers: res.headers, redirected };
}

export async function fetchHtml(url, opts = {}) {
  const doFetch = opts.fetch ?? undiciFetch;
  const maxBytes = opts.maxBytes ?? DEFAULT_MAX_BYTES;

  const { res, finalUrl, redirected } =
    opts.maxRedirects != null
      ? await followManually(doFetch, url, opts, opts.maxRedirects)
      : await followAuto(doFetch, url, opts);

  return finishFetch(res, finalUrl, redirected, opts, maxBytes);
}

File: rust/crates/turbo-crawl-core/src/net.rs
//! Networking (port of `src/net.mjs`, SPEC §8): `fetch_html` over reqwest
//! (redirects + gzip/br/deflate for free) plus the hardening the spec calls for:
//! charset sniffing, max body size, content-type gate, and an optional CookieJar
//! round-trip. The pure helpers below are unit-tested offline; `fetch_html`
//! itself is the live-IO seam (covered by the integration suite / harness).

use crate::cookies::CookieJar;
use bytes::BytesMut;
use futures_util::StreamExt;
use std::collections::BTreeMap;

const DEFAULT_UA: &str = "turbo-crawl/0.1 (+https://github.com/miaskiewicz/turbo-crawl)";
const DEFAULT_MAX_BYTES: usize = 8 * 1024 * 1024; // 8 MiB
const HTML_TYPES: &[&str] = &[
    "text/html",
    "application/xhtml+xml",
    "application/xml",
    "text/xml",
];

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum ErrorCode {
    BodyTooLarge,
    NotHtml,
    Network,
}

#[derive(Debug)]
pub struct HttpError {
    pub message: String,
    pub code: ErrorCode,
}

impl HttpError {
    fn new(message: impl Into<String>, code: ErrorCode) -> Self {
        Self {
            message: message.into(),
            code,
        }
    }
}

impl std::fmt::Display for HttpError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{:?}: {}", self.code, self.message)
    }
}
impl std::error::Error for HttpError {}

#[derive(Debug, Clone)]
pub struct FetchResult {
    pub html: String,
    pub final_url: String,
    pub status: u16,
    pub redirected: bool,
}

#[derive(Default)]
pub struct FetchOptions<'a> {
    pub headers: BTreeMap<String, String>,
    pub method: Option<String>,
    pub body: Option<String>,
    pub max_bytes: Option<usize>,
    pub allow_non_html: bool,
    pub max_redirects: Option<usize>,
    pub jar: Option<&'a mut CookieJar>,
    pub now: f64,
}

// charset: Content-Type header → <meta charset> sniff → utf-8.
fn detect_charset(content_type: &str, head: &[u8]) -> String {
    if let Some(cs) = charset_from_ct(content_type) {
        return cs;
    }
    charset_from_meta(head).unwrap_or_else(|| "utf-8".to_string())
}

fn charset_from_ct(ct: &str) -> Option<String> {
    let lc = ct.to_ascii_lowercase();
    let i = lc.find("charset=")? + "charset=".len();
    let rest = &lc[i..];
    let end = rest.find(';').unwrap_or(rest.len());
    Some(rest[..end].trim().replace(['"', '\''], ""))
}

fn charset_from_meta(head: &[u8]) -> Option<String> {
    let ascii: String = head
        .iter()
        .map(|&b| b as char)
        .collect::<String>()
        .to_ascii_lowercase();
    let i = ascii.find("charset=")? + "charset=".len();
    let label: String = ascii[i..]
        .chars()
        .skip_while(|c| *c == '"' || *c == '\'')
        .take_while(|c| c.is_ascii_alphanumeric() || *c == '-' || *c == '_')
        .collect();
    if label.is_empty() {
        None
    } else {
        Some(label)
    }
}

fn is_html_type(content_type: &str) -> bool {
    let ct = content_type.to_ascii_lowercase();
    if ct.is_empty() {
        return true; // permissive when the server omits it
    }
    HTML_TYPES.iter().any(|t| ct.contains(t))
}

fn decode(bytes: &[u8], charset: &str) -> String {
    let enc = encoding_rs::Encoding::for_label(charset.as_bytes()).unwrap_or(encoding_rs::UTF_8);
    enc.decode(bytes).0.into_owned()
}

// Method after a redirect: 303 (and 301/302 on POST) → GET; 307/308 keep it.
fn next_method(status: u16, method: &str) -> String {
    if status == 303 {
        return "GET".to_string();
    }
    if (status == 301 || status == 302) && method == "POST" {
        return "GET".to_string();
    }
    method.to_string()
}

const REDIRECT_STATUS: &[u16] = &[301, 302, 303, 307, 308];

fn is_redirect(status: u16) -> bool {
    REDIRECT_STATUS.contains(&status)
}

fn build_headers(url: &str, opts: &FetchOptions) -> BTreeMap<String, String> {
    let mut h = BTreeMap::new();
    h.insert("user-agent".into(), DEFAULT_UA.into());
    h.insert("accept".into(), "text/html,application/xhtml+xml".into());
    for (k, v) in &opts.headers {
        h.insert(k.to_ascii_lowercase(), v.clone());
    }
    if let Some(jar) = &opts.jar {
        let cookie = jar.cookie_header(url, opts.now);
        if !cookie.is_empty() {
            h.insert("cookie".into(), cookie);
        }
    }
    h
}

fn client(policy: reqwest::redirect::Policy) -> reqwest::Result<reqwest::Client> {
    reqwest::Client::builder().redirect(policy).build()
}

fn redirect_location(res: &reqwest::Response) -> Option<String> {
    if !is_redirect(res.status().as_u16()) {
        return None;
    }
    let loc = header_value(res, "location");
    if loc.is_empty() {
        None
    } else {
        Some(loc)
    }
}

fn err(e: impl std::fmt::Display, code: ErrorCode) -> HttpError {
    HttpError::new(e.to_string(), code)
}

async fn send(
    cl: &reqwest::Client,
    method: &str,
    url: &str,
    headers: &BTreeMap<String, String>,
    body: &Option<String>,
) -> Result<reqwest::Response, HttpError> {
    let m =
        reqwest::Method::from_bytes(method.as_bytes()).map_err(|e| err(e, ErrorCode::Network))?;
    let mut req = apply_headers(cl.request(m, url), headers);
    if let Some(b) = body {
        req = req.body(b.clone());
    }
    req.send().await.map_err(|e| err(e, ErrorCode::Network))
}

fn apply_headers(
    mut req: reqwest::RequestBuilder,
    headers: &BTreeMap<String, String>,
) -> reqwest::RequestBuilder {
    for (k, v) in headers {
        req = req.header(k.as_str(), v.as_str());
    }
    req
}

fn check_content_length(res: &reqwest::Response, max_bytes: usize) -> Result<(), HttpError> {
    if let Some(len) = res.content_length() {
        if len as usize > max_bytes {
            return Err(HttpError::new(
                format!("body exceeds maxBytes ({len} > {max_bytes})"),
                ErrorCode::BodyTooLarge,
            ));
        }
    }
    Ok(())
}

async fn read_capped(res: reqwest::Response, max_bytes: usize) -> Result<Vec<u8>, HttpError> {
    check_content_length(&res, max_bytes)?;
    let mut buf = BytesMut::new();
    let mut stream = res.bytes_stream();
    while let Some(chunk) = stream.next().await {
        let chunk = chunk.map_err(|e| HttpError::new(e.to_string(), ErrorCode::Network))?;
        if buf.len() + chunk.len() > max_bytes {
            return Err(HttpError::new(
                format!("body exceeds maxBytes (> {max_bytes})"),
                ErrorCode::BodyTooLarge,
            ));
        }
        buf.extend_from_slice(&chunk);
    }
    Ok(buf.to_vec())
}

fn header_value(res: &reqwest::Response, name: &str) -> String {
    res.headers()
        .get(name)
        .and_then(|v| v.to_str().ok())
        .unwrap_or("")
        .to_string()
}

fn ingest_set_cookie(opts: &mut FetchOptions, res: &reqwest::Response, final_url: &str) {
    let now = opts.now;
    let Some(jar) = opts.jar.as_mut() else { return };
    let lines: Vec<String> = res
        .headers()
        .get_all("set-cookie")
        .iter()
        .filter_map(|v| v.to_str().ok().map(str::to_string))
        .collect();
    if !lines.is_empty() {
        jar.set_from_response(final_url, &lines, now);
    }
}

fn gate_html_type(opts: &FetchOptions, res: &reqwest::Response) -> Result<(), HttpError> {
    let ct = header_value(res, "content-type");
    if !opts.allow_non_html && !is_html_type(&ct) {
        return Err(HttpError::new(
            format!("non-HTML content-type: {ct}"),
            ErrorCode::NotHtml,
        ));
    }
    Ok(())
}

// Turn a settled response into the FetchResult: gate content-type, then read +
// charset-decode the body under the byte cap.
async fn finish(
    opts: &FetchOptions<'_>,
    res: reqwest::Response,
    final_url: String,
    redirected: bool,
    max_bytes: usize,
) -> Result<FetchResult, HttpError> {
    gate_html_type(opts, &res)?;
    let status = res.status().as_u16();
    let ct = header_value(&res, "content-type");
    let bytes = read_capped(res, max_bytes).await?;
    let charset = detect_charset(&ct, &bytes[..bytes.len().min(1024)]);
    Ok(FetchResult {
        html: decode(&bytes, &charset),
        final_url,
        status,
        redirected,
    })
}

// State threaded through the manual-follow loop.
struct Hop {
    url: String,
    method: String,
    body: Option<String>,
}

// Manual redirect following (when `max_redirects` is set): re-derives the Cookie
// header and ingests Set-Cookie per hop and rewrites method/body per the fetch
// spec — the round-trip reqwest's auto-follow can't do.
async fn follow_manually(
    opts: &mut FetchOptions<'_>,
    start: &str,
    max_redirects: usize,
    max_bytes: usize,
) -> Result<FetchResult, HttpError> {
    let cl = client(reqwest::redirect::Policy::none()).map_err(|e| err(e, ErrorCode::Network))?;
    let mut hop = Hop {
        url: start.to_string(),
        method: opts.method.clone().unwrap_or_else(|| "GET".to_string()),
        body: opts.body.clone(),
    };
    for redirects in 0.. {
        let headers = build_headers(&hop.url, opts);
        let res = send(&cl, &hop.method, &hop.url, &headers, &hop.body).await?;
        let final_url = res.url().to_string();
        ingest_set_cookie(opts, &res, &final_url);
        match redirect_location(&res) {
            Some(loc) if redirects < max_redirects => hop = advance(&hop, &res, &loc)?,
            _ => return finish(opts, res, final_url, redirects > 0, max_bytes).await,
        }
    }
    unreachable!("redirect loop always returns at the cap")
}

fn advance(hop: &Hop, res: &reqwest::Response, loc: &str) -> Result<Hop, HttpError> {
    let status = res.status().as_u16();
    let method = next_method(status, &hop.method);
    let next = url::Url::parse(&hop.url)
        .and_then(|b| b.join(loc))
        .map_err(|e| err(e, ErrorCode::Network))?;
    let body = if method == hop.method {
        hop.body.clone()
    } else {
        None
    };
    Ok(Hop {
        url: next.to_string(),
        method,
        body,
    })
}

async fn follow_auto(
    opts: &mut FetchOptions<'_>,
    url: &str,
    max_bytes: usize,
) -> Result<FetchResult, HttpError> {
    let cl =
        client(reqwest::redirect::Policy::limited(20)).map_err(|e| err(e, ErrorCode::Network))?;
    let method = opts.method.clone().unwrap_or_else(|| "GET".to_string());
    let headers = build_headers(url, opts);
    let res = send(&cl, &method, url, &headers, &opts.body.clone()).await?;
    let final_url = res.url().to_string();
    let redirected = final_url != *url;
    ingest_set_cookie(opts, &res, &final_url);
    finish(opts, res, final_url, redirected, max_bytes).await
}

/// Fetch a URL and return its decoded HTML plus response metadata. With
/// `max_redirects` set, redirects are followed manually (per-hop cookie
/// round-trip, fetch-spec method rewrite); otherwise reqwest auto-follows
/// (cap 20). Content-type gate, byte cap, and charset decode mirror the JS.
pub async fn fetch_html(url: &str, mut opts: FetchOptions<'_>) -> Result<FetchResult, HttpError> {
    let max_bytes = opts.max_bytes.unwrap_or(DEFAULT_MAX_BYTES);
    match opts.max_redirects {
        Some(n) => follow_manually(&mut opts, url, n, max_bytes).await,
        None => follow_auto(&mut opts, url, max_bytes).await,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn charset_from_header() {
        assert_eq!(
            detect_charset("text/html; charset=ISO-8859-1", b""),
            "iso-8859-1"
        );
        assert_eq!(detect_charset("text/html; charset=\"utf-8\"", b""), "utf-8");
    }

    #[test]
    fn charset_from_meta_tag() {
        let head = br#"<html><head><meta charset="windows-1250">"#;
        assert_eq!(detect_charset("text/html", head), "windows-1250");
    }

    #[test]
    fn charset_falls_back_to_utf8() {
        assert_eq!(detect_charset("text/html", b"<html>"), "utf-8");
    }

    #[test]
    fn html_type_gate() {
        assert!(is_html_type(""));
        assert!(is_html_type("text/html; charset=utf-8"));
        assert!(!is_html_type("application/json"));
    }

    #[test]
    fn redirect_method_rewrite() {
        assert_eq!(next_method(303, "POST"), "GET");
        assert_eq!(next_method(301, "POST"), "GET");
        assert_eq!(next_method(307, "POST"), "POST");
        assert_eq!(next_method(302, "GET"), "GET");
    }

    #[test]
    fn redirect_status_set() {
        assert!(is_redirect(308));
        assert!(!is_redirect(200));
    }

    #[test]
    fn decode_latin1() {
        let bytes = [0xe9u8]; // é in latin1
        assert_eq!(decode(&bytes, "iso-8859-1"), "é");
    }
}
